### PR TITLE
Add alpine linux importer

### PR DESCRIPTION
**Pull request by sbs2001**
Signed-off-by: Shivam Sandbhor <shivam.sandbhor@gmail.com>

Fixes https://github.com/nexB/vulnerablecode/issues/43

I'm not sure what 'type' to assign to these packages (apk?)

### CODE REVIEW

**Comment by pombredanne** [comment]
I would use `alpine` as a package type

File: vulnerabilities/scraper/alpine_linux.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    vulnerability_package_dicts = []
+    for vulnerability in alpine_advisories(ALPINE_DB_URL):
+        for pkg_details in vulnerability['packages']:
+            package_name = pkg_details['pkg']['name']
+            for version, fixed_cves in pkg_details['pkg']['secfixes'].items():
+                # ['CVE-2016-9932  XSA-200', 'CVE-2016-9815','CVE-????-?????'] after mapping
+                # the split  function to above list
+                all_cves = list(map(lambda x: x.split(), fixed_cves))
+                # it becomes   [['CVE-2016-9932','XSA-200'], ['CVE-2016-9815'],['CVE-????-?????']]
+                all_cves = list(itertools.chain.from_iterable(all_cves))
```
Comment (haikoschol) [inline vulnerabilities/scraper/alpine_linux.py:28]: If you flatten the list, a non-CVE ID can get attributed to an unrelated vulnerability. For example with this data: https://github.com/alpinelinux/alpine-secdb/blob/master/v3.9/main.yaml#L1935

Comment (sbs2001) [reply vulnerabilities/scraper/alpine_linux.py:28]: Seeing this data, I see how wrong this approach, I will change this into something else.

File: vulnerabilities/management/commands/import.py [hunk 2]
```diff
@@ -24,9 +24,11 @@
 from django.core.management.base import BaseCommand, CommandError
 
 from vulnerabilities import data_dump as dd
-from vulnerabilities.scraper import debian, ubuntu, archlinux, npm, ruby, rust, safety_db
+from vulnerabilities.scraper import (
+        debian, ubuntu, archlinux, npm, ruby, rust, safety_db, alpine_linux)
 
 IMPORTERS = {
+    'alpinelinux': lambda: alpine_linux.import_vulnerabilities(),
```
[the remark this answers is not in the document]
Comment (sbs2001) [reply vulnerabilities/management/commands/import.py:31]: This is embarrassing. I will fix this.

[the remark this answers is not in the document]
Comment (haikoschol) [reply vulnerabilities/management/commands/import.py:31]: No worries. We should have tests for this. We don't so stuff like this is bound to happen.

File: vulnerabilities/data_dump.py [hunk 3]
```diff
@@ -299,3 +299,31 @@ def safetydb_dump(extract_data):
...
+            version=package_data['fixed_version']
+        )
+
+        for vulnerability in package_data['vuln_ids']:
+
+            if not vulnerability.startswith('CVE'):
+                # this links a non CVE to Vulnerability object of the
+                # previous CVE
+                VulnerabilityReference.objects.get_or_create(
+                    vulnerability=vulnerability_obj,
```
[the remark this answers is not in the document]
Comment (sbs2001) [reply vulnerabilities/data_dump.py]: I have made some changes in the importer, which includes making these checks, before passing the data to the `data_dump` process.

File: vulnerabilities/scraper/alpine_linux.py (hunk 1 above)
Comment (sbs2001) [reply vulnerabilities/scraper/alpine_linux.py]: Flattening the list was bad approach, please check my new approach

**Comment by haikoschol** [comment]
@sbs2001 as general feedback: please use a feature branch in your fork for future PRs instead of `develop`

**Comment by haikoschol** [comment]
@sbs2001 I tried to run the importer and it failed with the following error:
```
$ ./manage.py import alpinelinux
Importing data from alpinelinux
Traceback (most recent call last):
  File "./manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/home/hs/code/vulnerablecode/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/home/hs/code/vulnerablecode/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/hs/code/vulnerablecode/venv/lib/python3.8/site-packages/django/core/management/base.py", line 328, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/hs/code/vulnerablecode/venv/lib/python3.8/site-packages/django/core/management/base.py", line 369, in execute
    output = self.handle(*args, **options)
  File "/home/hs/code/vulnerablecode/vulnerabilities/management/commands/import.py", line 72, in handle
    self.import_data(sources)
  File "/home/hs/code/vulnerablecode/vulnerabilities/management/commands/import.py", line 87, in import_data
    IMPORTERS[src]()
  File "/home/hs/code/vulnerablecode/vulnerabilities/management/commands/import.py", line 31, in <lambda>
    'alpinelinux': lambda: dd.alpine_linux_dump(alpine_linux.import_vulnerabilities()),
  File "/home/hs/code/vulnerablecode/vulnerabilities/scraper/alpine_linux.py", line 21, in import_vulnerabilities
    for vulnerability in alpine_advisories(ALPINE_DB_URL):
  File "/home/hs/code/vulnerablecode/vulnerabilities/scraper/alpine_linux.py", line 12, in alpine_advisories
    with urlopen(url) as response:
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/usr/lib/python3.8/urllib/request.py", line 542, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 1362, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/usr/lib/python3.8/urllib/request.py", line 1319, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/lib/python3.8/http/client.py", line 1230, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1241, in _send_request
    self.putrequest(method, url, **skips)
  File "/usr/lib/python3.8/http/client.py", line 1092, in putrequest
    self._validate_path(url)
  File "/usr/lib/python3.8/http/client.py", line 1183, in _validate_path
    raise InvalidURL(f"URL can't contain control characters. {url!r} "
http.client.InvalidURL: URL can't contain control characters. '/alpine/infra/alpine-secdb/-/                archive/master/alpine-secdb-master.zip' (found at least ' ')
```

**Comment by sbs2001** [comment]
@haikoschol  can you check now? The error was caused because there was whitespace in the URL.

**Comment by haikoschol** [comment]
> @haikoschol can you check now? The error was caused because there was whitespace in the URL.

works

**Pull request merged**